### ISSUE TITLE
"'T' could be instantiated with an arbitrary type which could be unrelated" false positive when setter type is inferred

### ISSUE THREAD

**Issue by DetachHead** [microsoft/TypeScript#43492]
# Bug Report

### 🔎 Search Terms
- 'T' could be instantiated with an arbitrary type which could be unrelated
- setter
- generic

### 🕗 Version & Regression Information
4.3.0-dev.20210401

started happening since 4.3.0 beta

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.3.0-dev.20210401#code/MYGwhgzhAEBiD28A8AVAfNA3gKGn6A5gKYAu0AbmCAK5EAUAlAFzQpa76cBOp1XAdtACM0SNH5FyRLhzwBfWdAikKVWnUo0iDLHOgB6fZtrQAljFP8AZtJ4ATaCXjQARkWjwrjgJ4AHdwDkKAHYCsDw-BBkVojQALziRADucIhI-NQAtm5caIzY2DHwAHTG7glFpWruhkEB0OHUIA5uZpEkYPwkpmAkRA5JpiQAFqKCYFwuQ1wT3j7+0EnDpsCjjc2u7tT8POB9Dk7QARnZ0gHFBaCQMABCE6gYOJzEZGWMLGxPnPg8JHyCIjEEikMk4Ck4yle1Q01Q+OkwekMZTMMCIAA9fCAVkMQHNPPNAsFQthwu1XBN4okUncuOksjk8gwCi4JlUtJSWVw2SZDPxnLZ4FwgA)

### 💻 Code

```ts
class Foo<T> {
    get value(): T {
        return 1 as never
    }
    set value(value) {} //value is inferred to be of type 'T'
}
const foo = new Foo<number>()

foo.value = foo.value //'T' could be instantiated with an arbitrary type which could be unrelated to 'number'.

class Bar<T> {
    get value(): T {
        return 1 as never
    }
    set value(value: T) {} //value is explicitly of type 'T'
}
const bar = new Bar<number>()

bar.value = bar.value //no error
```

### 🙁 Actual behavior

`'T' could be instantiated with an arbitrary type which could be unrelated to 'number'.` error when using the `Foo.value` setter with the value from its getter

this doesn't happen when type is explicitly set to `T`

### 🙂 Expected behavior

no error, since as mentioned in the typescript 4.3 blog post, the type of the getter should always be assignable to the setter:

https://devblogs.microsoft.com/typescript/announcing-typescript-4-3-beta/#separate-write-types-on-properties

> One limitation of using different types for reading and writing properties is that the type for reading a property has to be assignable to the type that you’re writing. In other words, the getter type has to be assignable to the setter. This ensures some level of consistency, so that a property is always assignable to itself.
